### PR TITLE
Include ex.Data values in failure message (#3145)

### LINKED ISSUE #3145: Capture additional exception details in the test output
When our applications throw exceptions, we capture additional context in the .Data property of the exceptions. We would like to include this context in the NUnit output to reduce the time spent combing through logs and stepping through the debugger.

I've created two options for capturing additional information from exceptions.  https://github.com/drewburlingame/nunit/pull/1

> 1) Capture the exception in a new ExceptionResult.Exception property.
Pros:
    * A straightforward change. 
    * It can be inspected by components for different purposes.  
Con: 
    * Including the exception in the output of tests is more complicated, requiring an additional wrapper.  The output will not be included as part of the default NUnit error message

> 2) .  Register a static delegate with ExceptionHelper.GetAdditionalDetails
Pros:
    * A straightforward change. 
    * Very simple to use
    * Context is included with the exception message, making it easier to correlate.
Cons:
    * Doesn't fit with NUnit pattern of providing extensibility points via interfaces
    * Global to the app domain.  Cannot be enabled only for select fixtures.  This isn't an issue for our use case but I assume there are cases where this may be an issue.

Would either of these be acceptable solutions?  If not, do you have recommendations for how you'd like to see it?

### PR DESCRIPTION
**Pull request by drewburlingame**
Capture exception in ExceptionResult, enabling
extensions to perform more in-depth analysis &
reporting for failures

When value is null, replace with "<null>"

Closes #3145

### CODE REVIEW

File: src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                message = ExceptionHelper.BuildMessage(exception);
+            }
+            finally
+            {
+                ExceptionHelper.OutputExceptionDataProperty = true;
+            }
+            
+            Assert.That(message, !Contains.Substring("data-prop"));
+            Assert.That(message, !Contains.Substring("data-value"));
+        }
```
Comment (CharliePoole) [inline src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: I don't think we need to test turning it on and off for the same exception. The `finally` block is still needed because we changed a static property but as noted elsewhere I don't think we want to use a static.

I'd also like to see a test with a few values and have one test that checks the actual format of the message. The latter is more fragile, of course, but we should make sure the message is exactly as expected in one case anyway.

For completeness, let's have a test with nothing in the data dictionary. Of course that's tested by practically every other test we have, but it would be good to make it explicit here.

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs [hunk 2]
```diff
@@ -151,6 +166,24 @@ private static string GetExceptionMessage(Exception ex)
...
+            }
+
+            sb.AppendLine();
+            sb.AppendLine("Data:");
+            foreach (DictionaryEntry kvp in ex.Data)
+            {
+                sb.Append("  ");
+                sb.Append(kvp.Key);
+                sb.Append(": ");
+                sb.AppendLine(kvp.Value?.ToString() ?? "<null>");
```
[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I can change the appends.

[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I added "Data:" because it's valuable to know why the values are being output.  I'm working on another PR to support registering delegates to output additional details.  Section headers will be very helpful.

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs [hunk 3]
```diff
@@ -84,6 +91,10 @@ public static string BuildMessage(Exception exception, bool excludeExceptionName
             if (!excludeExceptionNames)
                 sb.AppendFormat("{0} : ", exception.GetType());
             sb.Append(GetExceptionMessage(exception));
+            if (OutputExceptionDataProperty)
+            {
+                AppendExceptionDataContents(exception, sb);
+            }
```
[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I'm not bothered either way.  If there's a standard for the project, I'm happy to follow it.

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs [hunk 4]
```diff
@@ -34,6 +36,11 @@ namespace NUnit.Framework.Internal
     /// </summary>
     public class ExceptionHelper
     {
+        /// <summary>
+        /// When true, each entry in Exception.Data is included in the exception message.
+        /// </summary>
+        public static bool OutputExceptionDataProperty { get; set; } = true;
```
[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: Sorry, I don't understand the first concern.  Are you saying that we want to include it by default and we shouldn't have this setting?

I was wondering about the parallel test runs.  I don't have a requirement for the feature to be optional so I'm happy to remove the setting if you think that's simpler all around.

File: src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs (hunk 1 above)
Comment (drewburlingame) [reply src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: if we make the feature not optional, this test can go away.

Comment (drewburlingame) [reply src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: Is the message format a contract?  Are there tests for the format I can follow?  I didn't see any.  Typically, I just verify message parts to reduce fragility.

Comment (drewburlingame) [reply src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: I can add an explicit test for the empty dictionary.  As you said, it's covered by all the other tests so I didn't think it necessary.  Explicit is good.

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I don't believe there is. Which leaves the choice up to the style of the individual, with others still free to grouse about it. :smile:

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I'm saying a few things here...

1. We need to decide if we want this to be optional as a team. If we don't, then you shouldn't add any sort of feature switch.

2. If we do want it, then we have to decide whether the switch should be for the entire run, the assembly, the fixture or the individual test - or some combo of these.

3. If we do want it to be optional, then it can't be a static member and therefore can't be a member of this static class. It has to be handled elssewhere.

4. If we do want it, then I'd rather see this issue deal with implementing it up to the point where it can be used. That's what I meant by saying no part of NUnit was using the switch. I'm not saying that your PR has to go that far, unless you want it to, another PR could actually finish up and close the issue.

[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: What's the process for making these decisions?  Is this PR blocked until the decision is made or is there something I can do to make the PR mergeable until the decision is made?

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: Life is boring when there's nothing to grouse about

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: We wait for feedback from other team members. Shouldn't take long.

Meanwhile, how do you think it should be resolved? If it's an option how do you think the user should express it?

[the remark this answers is not in the document]
Comment (drewburlingame) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: In regards to parallel test runs: I don't see users of NUnit changing the configuration during a test run, so static configuration would be fine in all cases... except NUnit test runs.  Does NUnit have any pattern for this?  If not, we could add the property to the test context and disable it within the test run.

In regards to developers configuring the settings: I imagined it could be done in the AssemblySetup.   If we add the setting to TestContext, then the static setting could be the default with overrides possible per test run.

In the next iteration, I had planned to add the ability for developers to register delegates for appending additional data to the message.  This way we could add a delegate to include the properties on the exception.  We could also add details for framework or corporate specific exceptions.  The way I would use it at my company is to reuse a component that prints information to logs for easier debugging.  It will be helpful to get the same information in our tests.  The component also outputs the Data element which is why it would be useful to disable this behavior in the framework.

[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: @drewburlingame 
We have used statics in that way in the past but have ended up regretting it a few times, when users wanted to change it for particular test cases. Use of a static limits testability and it's really not that hard to avoid it. I feel that the static has to go.

If we use the TestContext, it should not be a static but an instance setting or perhaps simply a known property of tests. An instance setting that is never changed passes down to all the child tests, so you get the stable behavior without the risk of using a static.

In keeping with the principle of separating policy (whether to include Data) from implementation (the hellper class) I would lean toward making `includeData` a parameter of the method that builds the message and putting the decision at a higher level.

@nunit/framework-team These are the options I see

1. No opt-in/opt-out switch

2. Opt-in globally, using an argument passed by the runner as a PackageSetting.

3. Opt-in at each level of the test hierarchy, as we do for default floating point tolerance.

I think all of these are valid ideas but we should really tell @drewburlingame which way we want to go so he can finish this up!

[the remark this answers is not in the document]
Comment (jnm2) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: I'm pretty much in agreement, but I'm leading towards 1.
Are exceptions with a non-empty Data member that common? The ResultStateExceptions we throw won't have anything in .Data, so it would have to be an uncaught exception in NUnit user code. Even then, .Data is typically empty. Then, adding the opt-in/opt-out would be just in case a user in this rare situation dislikes the Data readout enough that it they'd want the ability to stop seeing it. I'm thinking we can just wait until we're asked for an opt-out feature.

Potentially there's a scenario I don't know about. Does this sound reasonable?

**Comment by jnm2** [comment]
@drewburlingame By putting "Closes #3145" in your initial comment instead of in the PR title, you can have GitHub autoclose the issue when the PR is merged. (https://help.github.com/en/articles/closing-issues-using-keywords) Would you mind making that edit? Thanks!

File: src/NUnitFramework/framework/Internal/ExceptionHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitFramework/framework/Internal/ExceptionHelper.cs]: @jnm2 I think that makes sense. Let's not do something until we need it. This is what YAGNI is really all about: We know we will most likely need something, but we don't know the exact shape of what we'll need, so we leave it till we have more info.

**Comment by CharliePoole** [comment]
So, unless somebody else objects, I'd say take out the switch as well as making the other requested changes.

**Comment by jnm2** [comment]
ℹ @drewburlingame So whenever you get a chance again, we're ready for you to go ahead.

File: src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
+// ***********************************************************************
+// Copyright (c) 2019 Charlie Poole
```
Comment (rprouse) [inline src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: One copyright header and Charlie Poole, Rob Prouse 😄

Comment (drewburlingame) [reply src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: Fixed.  I have no idea how the duplicates happened since that was copy/paste.

Comment (jnm2) [reply src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs]: @rprouse @drewburlingame I bet this is the problem:

https://github.com/nunit/nunit/blob/edb0abc185af3e780e835bbb76318154e0b13091/NUnit.sln.DotSettings#L85-L105

**Comment by drewburlingame** [comment]
Is there anything else that needs to be done to merge this PR?

**Comment by CharliePoole** [comment]
@jnm2 requested two additional reviews, from me and @rprouse. I just noticed it myself and reviewed.
